File: java/smartlist/src/main/java/com/smartlist/module/organization/OrganizationService.java
package com.smartlist.module.organization;

import com.smartlist.model.AssetFile;
import com.smartlist.model.Organization;
import com.smartlist.model.User;
import com.smartlist.model.UserEmailInvite;
import com.smartlist.module.file.FileRepository;
import com.smartlist.module.organization.dto.request.OrganizationReqDTO;
import com.smartlist.module.organization.dto.request.InviteToOrganizationReqDTO;
import com.smartlist.module.organization.dto.request.OrganizationCancelInviteReqDTO;
import com.smartlist.module.organization.dto.response.OrganizationRespDTO;
import com.smartlist.module.organization.dto.response.OrganizationUserInviteListRespDTO;
import com.smartlist.module.organization.dto.response.OrganizationUserRespDTO;
import com.smartlist.module.user.UserEmailInviteRepository;
import com.smartlist.module.user.UserRepository;
import com.smartlist.services.CryptoService;
import com.smartlist.module.email.EmailService;
import org.apache.commons.lang3.ObjectUtils;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Service
public class OrganizationService {

    private final OrganizationRepository organizationRepository;
    private final FileRepository fileRepository;
    private final UserRepository userRepository;
    private final UserEmailInviteRepository userEmailInviteRepository;
    private final CryptoService cryptoService;
    private final EmailService emailService;
    private final ModelMapper modelMapper;

    public OrganizationService(final OrganizationRepository organizationRepository, final FileRepository fileRepository,
                               final UserRepository userRepository, final UserEmailInviteRepository userEmailInviteRepository,
                               final CryptoService cryptoService, final EmailService emailService,
                               final ModelMapper modelMapper){
        this.organizationRepository = organizationRepository;
        this.fileRepository = fileRepository;
        this.userRepository = userRepository;
        this.userEmailInviteRepository = userEmailInviteRepository;
        this.cryptoService = cryptoService;
        this.emailService = emailService;
        this.modelMapper = modelMapper;

        OrganizationReqDTO.addDTOToEntityMappings(this.modelMapper);
        OrganizationRespDTO.addEntityToDTOMappings(this.modelMapper);

    }


    public Integer createOrganization(final OrganizationReqDTO organizationReqDTO) {
        Optional<Organization> optionalOrganization  = this.organizationRepository.findByName(organizationReqDTO.getName());
        if(optionalOrganization.isPresent()){
            throw OrganizationResponseCodes.ORGANIZATION_NAME_EXISTS;
        }
        AssetFile logo = null;
        if (ObjectUtils.defaultIfNull(organizationReqDTO.getLogoId(), 0)  > 0) {
            Optional<AssetFile> optionalLogo = this.fileRepository.findById(organizationReqDTO.getLogoId());
            logo = optionalLogo.orElseThrow(() ->OrganizationResponseCodes.INVALID_LOGO_FILE_ID) ;
        }

        Organization organization = this.modelMapper.map(organizationReqDTO, Organization.class);
        organization.setLogo(logo);
        return this.organizationRepository.save(organization).getId();
    }

    public void updateOrganization(final Integer organizationId, final OrganizationReqDTO organizationReqDTO) {
        Optional<Organization> optionalOrganization = this.organizationRepository.findByName(organizationReqDTO.getName());

        if (optionalOrganization.isPresent() && !Objects.equals(optionalOrganization.get().getId(), organizationId)) {
            throw OrganizationResponseCodes.ORGANIZATION_NAME_EXISTS;
        }

        Organization organization = findOrganizationByIdOrThrow(organizationId);
        if (organization == null) {
            throw OrganizationResponseCodes.INVALID_ORGANIZATION_ID;
        }
        AssetFile logo = null;
        if (ObjectUtils.defaultIfNull(organizationReqDTO.getLogoId(), 0)  > 0) {
            Optional<AssetFile> optionalLogo = this.fileRepository.findById(organizationReqDTO.getLogoId());
            logo = optionalLogo.orElseThrow(() ->OrganizationResponseCodes.INVALID_LOGO_FILE_ID) ;
        }

        this.modelMapper.map(organizationReqDTO,organization);
        organization.setLogo(logo);
        this.organizationRepository.save(organization);
    }

    public OrganizationRespDTO getByOrganizationId(final Integer organizationId) {
        Optional<Organization> optionalOrganization = this.organizationRepository.findById(organizationId);
        if (optionalOrganization.isEmpty()) {
            throw OrganizationResponseCodes.INVALID_ORGANIZATION_ID;
        }

        return this.modelMapper.map(optionalOrganization.get(), OrganizationRespDTO.class);
    }

    public List<OrganizationRespDTO> getAllOrganizations() {
        List<Organization> organizations = this.organizationRepository.findAll();
        List<OrganizationRespDTO> organizationRespDTOs = new ArrayList<>();

        organizations.forEach(organization -> organizationRespDTOs.add(this.modelMapper.map(organization, OrganizationRespDTO.class)));
        return organizationRespDTOs;
    }

    public List<OrganizationUserRespDTO> getOrganizationUsers(final Integer organizationId) {
        Organization organization = findOrganizationByIdOrThrow(organizationId);
        ArrayList<OrganizationUserRespDTO> orgUsersRespDTOs = new ArrayList<>();

        List<User> users =  this.userRepository.findByOrganization(organization);
        users.forEach(user -> orgUsersRespDTOs.add(this.modelMapper.map(user, OrganizationUserRespDTO.class)));

        return orgUsersRespDTOs;
    }

    public void sendEmailInvites(final Integer organizationId, final InviteToOrganizationReqDTO inviteToOrganizationReqDto) {
        Organization organization = findOrganizationByIdOrThrow(organizationId);

        Optional<User> optionalUserInvitee  = this.userRepository.findById(inviteToOrganizationReqDto.getUserId());
        if (optionalUserInvitee.isEmpty()) {
            throw OrganizationResponseCodes.INVALID_USER_ID;
        }

        for (int i = 0; i < inviteToOrganizationReqDto.getEmails().size(); i++) {
            String emailId = inviteToOrganizationReqDto.getEmails().get(i);
            Optional<User> optionalUser = this.userRepository.findByEmail(emailId);
            if (optionalUser.isPresent()) {
                continue;
            }
            Optional<UserEmailInvite> optionalExistingUserEmailInvite  = this.userEmailInviteRepository.findByEmail(emailId);
            if (optionalExistingUserEmailInvite.isPresent()) {
                continue;
            }

            UserEmailInvite userEmailInvite = new UserEmailInvite();
            userEmailInvite.setEmail(emailId);
            userEmailInvite.setOrganization(organization);
            userEmailInvite.setToken(this.cryptoService.generateToken());
            this.userEmailInviteRepository.save(userEmailInvite);

            this.emailService.sendInviteToAppMail(emailId, optionalUserInvitee.get().getName(), userEmailInvite.getToken());
        }
    }

    public OrganizationUserInviteListRespDTO getUserInvites(final Integer organizationId){
        Organization organization = findOrganizationByIdOrThrow(organizationId);
        List<UserEmailInvite> userEmailInvites = this.userEmailInviteRepository.findByOrganization(organization);

        return new OrganizationUserInviteListRespDTO(userEmailInvites);
    }

    public void cancelInviteToOrganization(final Integer organizationId, final OrganizationCancelInviteReqDTO cancelInviteToOrganizationReqDto) {
        Organization organization = findOrganizationByIdOrThrow(organizationId);
        Optional<UserEmailInvite> optionalUserEmailInvite = this.userEmailInviteRepository.findByEmailAndOrganization(cancelInviteToOrganizationReqDto.getEmail(), organization);
        if (optionalUserEmailInvite.isEmpty()) {
            throw OrganizationResponseCodes.INVALID_USER_EMAIL_ID;
        }
        this.userEmailInviteRepository.delete(optionalUserEmailInvite.get());
    }


    private Organization findOrganizationByIdOrThrow(final Integer organizationId) {
        Optional<Organization> optionalOrganization = this.organizationRepository.findById(organizationId);
        return optionalOrganization.orElseThrow (() -> OrganizationResponseCodes.INVALID_ORGANIZATION_ID);
    }
}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/organization/organization.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { plainToClass } from 'class-transformer';
import { Organization } from 'src/entities/organization.entity';
import { User } from 'src/entities/user.entity';
import { AssetFile } from 'src/entities/asset-file.entity';
import { Repository } from 'typeorm';
import { CreateOrganizationReqDto } from './dto/request/create-organization.req.dto';
import { InviteToOrganizationReqDto } from './dto/request/invite-to-organization.req.dto';
import { UpdateOrganizationReqDto } from './dto/request/update-organization.req.dto';
import { OrganizationUserRespDto } from './dto/response/organization-user.resp.dto';
import { OrganizationRespDto } from './dto/response/organization.resp.dto';
import { OrganizationRepository } from './organization.repository';
import { OrganizationCancelInviteReqDto } from './dto/request/cancel-invite-to-organization.req.dto';
import { UserEmailInvite } from 'src/entities/user-email-invite.entity';
import { OrganizationUserInviteListRespDto } from './dto/response/organization-user-invite-list.req.dto';
import { OrganizationResponseCodes } from './organization.response.codes';
import { CryptoService } from 'src/providers/crypto.service';
import { EmailService } from 'src/providers/email.service';
import { FileRepository } from '../file/file.repository';

@Injectable()
export class OrganizationService {
  constructor(
    @InjectRepository(OrganizationRepository)
    private organizationRepository: OrganizationRepository,
    @InjectRepository(User)
    private userRepository: Repository<User>,
    @InjectRepository(UserEmailInvite)
    private userEmailInviteRepository: Repository<UserEmailInvite>,
    @InjectRepository(FileRepository)
    private fileRepository: FileRepository,
    private cryptoService: CryptoService,
    private emailService: EmailService,
  ) {}

  async createOrganization(
    createOrganizationDto: CreateOrganizationReqDto,
  ): Promise<number> {
    let organization: Organization = await this.organizationRepository.findOne({
      name: createOrganizationDto.name,
    });

    if (organization != null) {
      throw OrganizationResponseCodes.ORGANIZATION_NAME_EXISTS;
    }
    let logo: AssetFile;
    if (createOrganizationDto.logo_id > 0) {
      logo = await this.fileRepository.findOne(createOrganizationDto.logo_id);
      if (logo == null) {
        throw OrganizationResponseCodes.INVALID__LOGO_FILE_ID;
      }
    }

    organization = plainToClass(Organization, createOrganizationDto);
    organization.logo = logo;
    organization = await this.organizationRepository.save(organization);
    return organization.id;
  }

  async updateOrganization(
    id: number,
    updateOrganizationReqDto: UpdateOrganizationReqDto,
  ) {
    let organization: Organization = await this.organizationRepository.findOne({
      name: updateOrganizationReqDto.name,
    });

    if (organization != null && organization.id != id) {
      throw OrganizationResponseCodes.ORGANIZATION_NAME_EXISTS;
    }

    organization = await this.findOrganizationByIdOrThrow(id);
    if (organization == null) {
      throw OrganizationResponseCodes.INVALID_ORGANIZATION_ID;
    }

    let logo: AssetFile = null;
    if (updateOrganizationReqDto.logo_id > 0) {
      logo = await this.fileRepository.findOne(
        updateOrganizationReqDto.logo_id,
      );
      if (logo == null) {
        throw OrganizationResponseCodes.INVALID__LOGO_FILE_ID;
      }
    }

    organization = plainToClass(Organization, updateOrganizationReqDto);
    organization.id = id;
    organization.logo = logo;

    await this.organizationRepository.save(organization);
  }

  async sendEmailInvites(
    id: number,
    inviteToOrganizationReqDto: InviteToOrganizationReqDto,
  ) {
    const organization: Organization = await this.findOrganizationByIdOrThrow(
      id,
    );

    const userInvitee: User = await this.userRepository.findOne(
      inviteToOrganizationReqDto.userId,
    );
    if (userInvitee == null) {
      throw OrganizationResponseCodes.INVALID_USER_ID;
    }

    for (let i = 0; i < inviteToOrganizationReqDto.emails.length; i++) {
      const emailId = inviteToOrganizationReqDto.emails[i];
      const user: User = await this.userRepository.findOne({
        email: emailId,
      });
      if (user != null) {
        continue;
      }
      const exisitingUserEmailInvite: UserEmailInvite =
        await this.userEmailInviteRepository.findOne({
          email: emailId,
        });
      if (exisitingUserEmailInvite != null) {
        continue;
      }

      const userEmailInvite: UserEmailInvite = new UserEmailInvite();
      userEmailInvite.email = emailId;
      userEmailInvite.organization = organization;
      userEmailInvite.token = this.cryptoService.generateToken();
      await this.userEmailInviteRepository.save(userEmailInvite);
      this.emailService.sendInviteToAppMail(
        emailId,
        userInvitee.name,
        userEmailInvite.token,
      );
    }
  }

  async getByOrganizationId(id: number): Promise<OrganizationRespDto> {
    const organization: Organization =
      await this.organizationRepository.findByIdWithRelations(id);

    if (organization == null) {
      throw OrganizationResponseCodes.INVALID_ORGANIZATION_ID;
    }

    return new OrganizationRespDto(organization);
  }

  async getAllOrganizations(): Promise<OrganizationRespDto[]> {
    const organizations: Organization[] =
      await this.organizationRepository.findWithRelations();
    const organizationRespDto: OrganizationRespDto[] = [];
    organizations.forEach((organization) => {
      organizationRespDto.push(new OrganizationRespDto(organization));
    });
    return organizationRespDto;
  }

  async getUsers(id: number): Promise<OrganizationUserRespDto[]> {
    const org: Organization = await this.findOrganizationByIdOrThrow(id);
    const orgUsersRespDtoArray = new Array<OrganizationUserRespDto>();

    const users: User[] = await this.userRepository.find({ organization: org });
    users.forEach(function (user) {
      orgUsersRespDtoArray.push(new OrganizationUserRespDto(user));
    });

    return orgUsersRespDtoArray;
  }

  async getUserInvites(id: number) {
    const org: Organization = await this.findOrganizationByIdOrThrow(id);
    const userEmailInvites: UserEmailInvite[] =
      await this.userEmailInviteRepository.find({
        organization: org,
      });

    return new OrganizationUserInviteListRespDto(userEmailInvites);
  }

  async cancelInviteToOrganization(
    id: number,
    cancelInviteToOrganizationReqDto: OrganizationCancelInviteReqDto,
  ) {
    const org: Organization = await this.findOrganizationByIdOrThrow(id);
    const userEmailInvite: UserEmailInvite =
      await this.userEmailInviteRepository.findOne({
        email: cancelInviteToOrganizationReqDto.email,
        organization: org,
      });
    if (userEmailInvite == null) {
      throw OrganizationResponseCodes.INVALID_USER_EMAIL_ID;
    }
    await this.userEmailInviteRepository.remove(userEmailInvite);
  }

  private async findOrganizationByIdOrThrow(
    organizationId: number,
  ): Promise<Organization> {
    const organization: Organization =
      await this.organizationRepository.findOne(organizationId);
    if (organization == null) {
      throw OrganizationResponseCodes.INVALID_ORGANIZATION_ID;
    }
    return organization;
  }
}
